### <p style="text-align: right;"> &#9989; Nate Kurt
<p style="text-align: right;"> &#9989; Patrick Govan, Jack M, Noah Z, Matt J, Vince P

# Day 3 In-class Assignment: The perfectly timed snowball
![Snowballs](http://cocktailsandmovies.com/wp-content/uploads/2015/12/Elf-snowballs.jpg)

## Learning Goals:

By the end of this assignment you should be able to:

* Use Python lists to store values
* Use loops to iterate over a specified range or until a condition is met.
* Compute the speed necessary to throw a perfectly timed snowball.


---
## The Problem

Given all the snow that will arrive sooner than we'd like here in Michigan, you've been thinking about playing a little prank on your roommate this winter involving a perfectly timed snowball. Fortunately, you've been learning a bit of Python and you're pretty sure that you can write a bit of code to make sure your plan goes without a hitch.

### The details of your plan

You plan to wait at home for your roommate to get back from class. You figure you can hideout on the porch and when you see your roommate coming, you can throw the snowball straight up in the air and duck inside so that you can watch from the window as they receive your icy surprise.

This sounds like a great plan, but you realize you need to know when the snowball will land based on how fast you throw it so that you can time it just right. You're not yet sure how long it takes your roommate to get to the door from the moment when you first see them, so you figure you'll just compute several impacts times based on a range of initial snowball speeds and work out the rest of the details later.

### The math behind the model

Now that you've come up with your plan, you need to do a bit of digging to figure out how to compute the amount of time the snowball spends in the air based on the speed you launch it with. You spend a bit of time searching around on the internet, including Wikipedia, and you find that the important equation of motion is this:

$$h(t) = h_0 + v_0t - \frac{1}{2}gt^2$$

where $h$ is the height of the snowball as a function of time, $h_0$ is the initial height that the snowball is launched from, $v_0$ is the initial speed the snowball is launched with, $g$ is the acceleration due to gravity, and $t$ is the time in seconds. 

You give it some further thought and realize what you want to know is when will the snowball land, so **when does $h(t)=0$?**

You know that the initial height that the snowball is launched from, $h_0$, is roughly your height since you'll be throwing it, which we'll approximate as **1.75 meters**.

You also recall that at some point you learned that the **acceleration due to gravity is 9.8 m/s$^2$**

This means that all you need to do is **calculate the value of $t$ based on a range of initial values for the speed of the snowball, $v_0$**.

#### Potentially useful information:

It might have been a while since you've see an equation that looks like the one above. It turns out it's very close to being a **quadratic equation**, which looks something like:

$$at^2 + bt + c = 0$$

and, if you recall, the solution to this equation is given by:

$$t = \frac{-b~\pm~\sqrt{b^2~-~4ac}}{2a}$$

---
## Part 1: Planning your solution and pseudo-coding

In general, before you start writing any code, it's good to get into the habit of coming up with a plan for how you've going to solve the problem. During this phase it is important to identify and define the variables that are revelant to the problem, the equations you're going to need to solve, and the overall structure of the program that you will write to solve the problem.

To get some practice with this, you and your group members are expected to **use your whiteboards** to come up with a plan of attack for computing the impact time of the snowball you're hoping to drop on the unsuspecting head of your roommate. Part of this planning process should involve a bit of what we commonly refer to as "pseudo-code." Pseudo-code is writing or text that follows the programmatic style of real code but doesn't necessarily have all of the proper syntax. You can think of it as an outline for how your program *should* work, before you actually start to code it up.

Things you should decide on during your planning and psuedo-coding:

1. What variables are important for solving this problem?
2. What are the right initial values for these variables?
3. What is a reasonable range for the snowball speeds you're going to explore?
4. How will you go about solving for multiple impact times?
5. How many impact times do you plan to compute?
6. How will you store your snowball speeds and corresponding impacts times?

Not only should you come up with answers to these questions and a plan for your code, but you should be able to defend your choices!

**Record the details of you plan in the Markdown cell below.**

1. velocity inital and final,  and time, 
2. time = 0, vinit = 10m/s to 20m/s, 
3. range 10m/s to 20m/s 
4. multiple impact times just adjust inital velocity


---
### STOP
Check-in with an instructor before you start coding!

---

---
## Part 2: Coding up your solution

Now that you and your group have come up with a plan for your code, it's time to put your plan into action. Your code should do the following:

1. Compute snowball impact times for a range of initial snowball speeds. You are free to choose the range of initial snowball speeds (the number of speeds and the value of each speed), but you are expected to justify your choices.
2. Store the snowball speeds and their corresponding impact times in a list.
3. Iterate through your list(s) of speeds and impact times and print the values in the follow way:
    - "For an initial snowball speed of < v > m/s, the impact time will be < t > seconds." where < v > is the speed and < t > is the time
    
**Write your code in the cell below (feel free to break it up into multiple cells if you prefer).**

In [18]:
# Put your code for computing snowball impact times here!
def quadratic(a,b,c):
    pos = (-1 * b + ((b **2) - (4* a * c)) ** (1/2)) / (2 * a)
    neg = (-1 * b - ((b **2) - (4* a * c)) ** (1/2)) / (2 * a)
    return (round(neg,3))

init_height = 1.75
gravity = (9.8) * (-.5)
speed = [10,11,12,13,14,15,16,17,18,19,20]



for i in speed:
    print("When initial speed is %s m/s: " % (i), quadratic(gravity,i,init_height))


speed_list = [quadratic(gravity,i,init_height) for i in speed]

print(speed_list)

new_list = list(zip(speed,speed_list))


    
    

When initial speed is 10 m/s:  2.203
When initial speed is 11 m/s:  2.394
When initial speed is 12 m/s:  2.587
When initial speed is 13 m/s:  2.781
When initial speed is 14 m/s:  2.977
When initial speed is 15 m/s:  3.174
When initial speed is 16 m/s:  3.371
When initial speed is 17 m/s:  3.569
When initial speed is 18 m/s:  3.768
When initial speed is 19 m/s:  3.968
When initial speed is 20 m/s:  4.167


**Question 1**: What did you choose as your slowest speed and why? What is the corresponding impact time for this speed?

0 would be the slowest speed because at the top of the parabola 

**Question 2**: What did you choose as your fastest speed and why? What is the corresponding impact time for this speed?

Fastest speed was 20 m/s and time was 4.167

**Question 3**: Based on the impact times that you computed, do you think it will be feasible to pull off the prank without your roommate catching you? Explain why or why not.

Not really, you'd have to be pretty accurate and also pretty fast to get away since it happens all in like four seconds 

**Question 4**: What was the hardest part about computing the impact times for your snowball prank?

I guess converting the values

---
### STOP
Check in with an instructor before you move on!

---

---
## Part 3: Finding the maximal height

Now that you have an idea how long the snowball will be in the air based on how hard you throw it, you also want to find out how high the snowball will travel in the air before falling towards the ground. Similarly to Part 1, you should write some __pseudo-code__ that will act as a guide for how you _would_ determine:

1. The maximal height of the snowball based on its initial velocity.
2. The time at which the snowball achieves the maximal height.

You should discuss how you would determine these values with your group members and make use of the whiteboards to help map out the basic ideas. Your solution might require some concepts that we haven’t yet learned — that’s OK! You just need to figure out how the code _would_ work.
 
**Record the details of you plan in the Markdown cell below.**

Take derivative of equation and then find value at 0

so v(t) = V0 - gt
so 0 = V0 - (9.8)(t)
   V0/9.8 = t
   
 then plug in that value for the height equation for t

---

## Time Allowing: 
Try implementing your solution to Part 3 with code. If you get stuck on a certain concept, ask the instructor, TA, or LA(s) for help.

In [38]:
# Put your code for computing the maximal height and when the snowball achieves this maximal height here!
def max_height(v):
    time = v/9.8
    height = 1.75 + (v * time) - ((4.9) * time**2)
    return round(height, 3)

def max_height_time(v):
    return v/9.8

speed = [10,11,12,13,14,15,16,17,18,19,20]
for i in speed:
    print("At time %s seconds, max height is %sm" % (round(max_height_time(i),3), max_height(i)))
    

At time 1.02 seconds, max height is 6.852m
At time 1.122 seconds, max height is 7.923m
At time 1.224 seconds, max height is 9.097m
At time 1.327 seconds, max height is 10.372m
At time 1.429 seconds, max height is 11.75m
At time 1.531 seconds, max height is 13.23m
At time 1.633 seconds, max height is 14.811m
At time 1.735 seconds, max height is 16.495m
At time 1.837 seconds, max height is 18.281m
At time 1.939 seconds, max height is 20.168m
At time 2.041 seconds, max height is 22.158m


---
### STOP
Check in with an instructor before you leave class!

---

---
## Assignment Wrap-up

Fill out the following Google form before submitting your notebook to D2L.

In [39]:
from IPython.display import HTML

HTML("""
<iframe src="https://goo.gl/forms/C4A2pmcc3k2HuGjm2"
width="1000"
height="1000"
frameborder="0"
marginheight="0"
marginwidth="0"
>Loading...</iframe>
""")

## Congratulations, you're done!

Submit this assignment by uploading it to the course Desire2Learn web page.  Go to the "In-class assignments" folder, find the submission link for Day 3, and upload it there.

See you next class!

&#169; Copyright 2018,  Michigan State University Board of Trustees